### PR TITLE
Add Vault contracts into new project structure

### PR DESCRIPTION
**Pull request by alistair-singh**
TODO:
- [x] e2e tests
- [x] Add more tests

### CODE REVIEW

**Review by claravanstaden — APPROVED** [review]

File: core/packages/contracts/contracts/ERC20Vault.sol [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    /// @param recipient The address of the sender.
+    /// @param token The address of the ERC20 token.
+    /// @param amount The amount being withdrawn.
+    event Withdraw(address account, address recipient, address token, uint128 amount);
+
+    /// @dev Token Transfer failed.
+    error TokenTransferFailed();
+
+    /// @dev Not enough funds to transfer.
+    error InsufficientBalance();
```
[the remark this answers is not in the document]
Comment (alistair-singh) [reply core/packages/contracts/contracts/ERC20Vault.sol:31]: This for raising this. It seems we do this check in EthApp. 

https://github.com/Snowfork/snowbridge/pull/722/files#diff-183a0a4e6dac52a570de98816f9f67063c5ccb8b51fe33895fd3ade207e8c63dR102

However its  just named `ExceedsBalance`. I have changed both the errors to be `InsufficientBalance` and will handle them in the vault.

File: core/packages/contracts/contracts/DOTApp.sol [hunk 2]
```diff
@@ -68,4 +68,13 @@ contract DOTApp is FeeController, AccessControl {
     {
         token.burn(feePayer, _amount);
     }
+
+    function transferVaultOwnership(
```
Comment (vgeddes) [inline core/packages/contracts/contracts/DOTApp.sol:72]: Lets rather implement this after the XCMv3 work as we're likely going to have rewrite a lot of the app code.

File: core/packages/contracts/contracts/ERC20App.sol [hunk 3]
```diff
@@ -109,13 +96,16 @@ contract ERC20App is AccessControl {
...
-        }
-
-        balances[_token] = balances[_token] - _amount;
-        IERC20(_token).safeTransfer(_recipient, _amount);
-
+        vault.withdraw(_recipient, _token, _amount);
         emit Unlocked(_token, _sender, _recipient, _amount);
     }
+
+    function transferVaultOwnership(
```
Comment (vgeddes) [inline core/packages/contracts/contracts/ERC20App.sol:103]: ditto. Don't want to touch governance/upgrade stuff until we have a sense on how XCMv3 will change things.

File: core/packages/contracts/contracts/EtherVault.sol [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            revert InsufficientBalance();
+        }
+        require(_amount > 0, "Must unlock a positive amount");
+        (bool success, ) = _recipient.call{ value: _amount }("");
+        if(!success) {
+            revert CannotWithdraw();
+        }
+        emit Withdraw(msg.sender, _recipient, _amount);
+    }
+}
```
Comment (vgeddes) [inline core/packages/contracts/contracts/EtherVault.sol:60]: Missing newline at end of file. If you're using VS code I highly recommend enabling this config option: `files.insertFinalNewline`

**Review by vgeddes — APPROVED** [review]

**Pull request merged**